dask_kubernetes/operator/controller/controller.py

```diff
@@ -392,67 +411,92 @@ async def get_desired_workers(scheduler_service_name, namespace, logger):
         ) from e
 
 
-@kopf.on.update("daskworkergroup.kubernetes.dask.org")
-async def daskworkergroup_update(spec, name, namespace, logger, **kwargs):
-    async with kubernetes.client.api_client.ApiClient() as api_client:
-        customobjectsapi = kubernetes.client.CustomObjectsApi(api_client)
-        corev1api = kubernetes.client.CoreV1Api(api_client)
+worker_group_scale_locks = defaultdict(lambda: asyncio.Lock())
 
-        cluster = await customobjectsapi.get_namespaced_custom_object(
-            group="kubernetes.dask.org",
-            version="v1",
-            plural="daskclusters",
-            namespace=namespace,
-            name=spec["cluster"],
-        )
-        cluster_labels = cluster.get("metadata", {}).get("labels", {})
 
-        workers = await corev1api.list_namespaced_pod(
-            namespace=namespace,
-            label_selector=f"dask.org/workergroup-name={name}",
-        )
-        current_workers = len(
-            [w for w in workers.items if w.status.phase != "Terminating"]
-        )
-        desired_workers = spec["worker"]["replicas"]
-        workers_needed = desired_workers - current_workers
-        annotations = _get_dask_cluster_annotations(kwargs["meta"])
-        if workers_needed > 0:
-            for _ in range(workers_needed):
-                data = build_worker_pod_spec(
-                    worker_group_name=name,
-                    namespace=namespace,
-                    cluster_name=spec["cluster"],
-                    uuid=uuid4().hex[:10],
-                    spec=spec["worker"]["spec"],
-                    annotations=annotations,
-                )
-                kopf.adopt(data)
-                kopf.label(data, labels=cluster_labels)
-                await corev1api.create_namespaced_pod(
+@kopf.on.field("daskworkergroup.kubernetes.dask.org", field="spec.worker.replicas")
+async def daskworkergroup_replica_update(
+    name, namespace, meta, spec, new, body, logger, **kwargs
+):
+    cluster_name = spec["cluster"]
+
+    # Replica updates can come in quick succession and the changes must be applied atomically to ensure
+    # the number of workers ends in the correct state
+    async with worker_group_scale_locks[f"{namespace}/{name}"]:
```

Comment:
I wonder if we should generate some UUID constant at the top of the file. Then to take the lock we set an annotation on the `DaskWorkerGroup` like `kubernetes.dask.org/scale-in-progress` and set the value to that UUID.

Then as part of trying to take the lock we check for that annotation. If it exists and the UUID matches we know the current controller is busy scaling in another coroutine, so we wait. If the UUID doesn't match we can assume the controller was restarted or failed over and the in-progress scale has been aborted. So we can update the annotation with our UUID and take the lock.

Then at the end, we delete the annotation again to release the lock.

This functionality could be factored into a small async context manager so that the code here stays basically the same.

Reply:
This generally sounds good to me - I've responded to this in the other comment as both comments are related.